src/Framework/FileClassifier.cs

```diff
@@ -0,0 +1,196 @@
+        private static readonly char[] s_semicolonDelimiter = {';'};
+
+        /// <summary>
+        ///     Singe, static <see cref="Lazy{T}"/> instance of shared file FileClassifier for <see cref="Shared"/> member.
+        /// </summary>
+        private static readonly Lazy<FileClassifier> s_sharedInstance = new(() => new FileClassifier());
+
+        /// <summary>
+        ///     Serves purpose of thread safe set of known immutable directories.
+        /// </summary>
+        /// <remarks>
+        ///     Although <see cref="ConcurrentDictionary{TKey,TValue}"></see> is not memory wise optimal solution, int this particular case it does no matter
+        ///     as much as expected size of this set is ~5 and in very extreme cases less then 100.
+        /// </remarks>
+        private readonly ConcurrentDictionary<string, string> _knownImmutableDirectories = new();
+
+        /// <summary>
+        ///     Copy on write snapshot of <see cref="_knownImmutableDirectories"/>.
+        /// </summary>
+        private IReadOnlyList<string> _knownImmutableDirectoriesSnapshot = Array.Empty<string>();
+
+        /// <summary>
+        ///     Creates default FileClassifier which following immutable folders:
+        ///     Classifications provided are:
+        ///     <list type="number">
+        ///         <item>Program Files</item>
+        ///         <item>Program Files (x86)</item>
+        ///         <item>Default .nuget cache location</item>
+        ///         <item>Visual Studio installation root</item>
+        ///     </list>
+        /// </summary>
+        /// <remarks>
+        ///     Individual projects NuGet folders are added during project build by calling
+        ///     <see cref="RegisterImmutableDirectories" />
+        /// </remarks>
+        public FileClassifier()
+        {
+            string? programFiles32 = Environment.GetEnvironmentVariable("ProgramFiles(x86)");
+            string? programFiles64 = Environment.GetEnvironmentVariable("ProgramW6432");
+
+            if (!string.IsNullOrEmpty(programFiles32))
+            {
+                RegisterImmutableDirectory(Path.Combine(programFiles32, "Reference Assemblies", "Microsoft"));
+                RegisterImmutableDirectory(Path.Combine(programFiles32, "dotnet"));
+            }
+            if (!string.IsNullOrEmpty(programFiles64))
+            {
+                RegisterImmutableDirectory(Path.Combine(programFiles64, "Reference Assemblies", "Microsoft"));
+                RegisterImmutableDirectory(Path.Combine(programFiles64, "dotnet"));
+            }
+            RegisterImmutableDirectory(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget", "packages"));
+            RegisterImmutableDirectory(GetVSInstallationDirectory());
+
+            return;
+
+            static string? GetVSInstallationDirectory()
+            {
+                string? dir = Environment.GetEnvironmentVariable("VSAPPIDDIR");
```

Comment:
I don't think this variable is reliable; it's not defined on my machine for instance:

```sh-session
❯ ls env:VS*

Name                           Value
----                           -----
VSSDKINSTALL                   C:\Program Files\Microsoft Visual Studio\2022\Preview\VSSDK
VS170COMNTOOLS                 C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\Tools\
VSCMD_VER                      17.1.0-pre.2.0
VSCMD_ARG_HOST_ARCH            x86
VSSDK150INSTALL                C:\Program Files\Microsoft Visual Studio\2022\Preview\VSSDK
VSCMD_ARG_TGT_ARCH             x86
VSCMD_ARG_app_plat             Desktop
VSINSTALLDIR                   C:\Program Files\Microsoft Visual Studio\2022\Preview\
```

Can you use something like the current-assembly-directory-minus-some-layers code in BuildEnvironmentHelper? And yeah, please add a comment about disentangling from shared (on that glorious future day).

https://github.com/dotnet/msbuild/blob/24b33188f385cee07804cc63ec805216b3f8b72f/src/Shared/BuildEnvironmentHelper.cs#L317-L324

Reply:
VSAPPIDDIR env var is created and VS process. So the code will always work if MSBuild is invoked/spawned from VS.
If MSBuild is invoked from command line this will not work. 
However, according to our traces there is only limited amount of files with multiply up-to-date checks from VS folder. I am just trying to do the best effort here, maybe we can consider to not include VS in there for now. Shall we?

Reply:
> However, according to our traces there is only limited amount of files with multiply up-to-date checks from VS folder.

Really? I'd expect them to be quite common since there are a bunch of `.targets` files installed with VS and we should be checking ~all of them at every project load to get the latest one for `$(MSBuildAllProjects)`.

Reply:
Hmmm, you are right. I have been misled by the fact that I have captured traces when running MSBuilld from bootstrap artifacts. Now when I look at it, I have find that bellow files are queried quite often:
file | # of checks
-- | --
E:\dev\msbuild\artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\amd64\Microsoft.Common.targets | 149
E:\dev\msbuild\artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\amd64\Microsoft.Xaml.targets | 504

However, to use BuildEnvironmentHelper.cs i'd have to include all bunch of Shared files in Framework. I can of course clone the related parts of code. What would you recommend?

Reply:
I'm fine with copy/paste and a note.

Reply:
Done. Quite a few lines has been cloned.